packages/typescript/src/internal/ts_project.bzl

```diff
@@ -0,0 +1,192 @@
+    - Bazel expects that each output is produced by a single rule.
+      Thus if you have two `ts_project` rules with overlapping sources (the same .ts file
+      appears in more than one) then you get an error if you try to build both together.
+      Worse, if you build them separately then the output directory will contain whichever
+      one you happened to build most recently. This is highly discouraged.
+
+    > Note, in order for TypeScript to find referenced projects in the bazel-out folder,
+    > we recommend that the base tsconfig contain a rootDirs section that includes all
+    > possible locations they may appear.
+    >
+    > We hope this will not be needed in some future release of TypeScript.
+    > Follow https://github.com/microsoft/TypeScript/issues/37257 for more info.
+    >
+    > For example, if the base tsconfig file relative to the workspace root is
+    > `path/to/tsconfig.json` then you should configure like:
+
+    ```
+    "compilerOptions": {
+        "rootDirs": [
+            ".",
+            "../../bazel-out/darwin-fastbuild/bin/path/to",
+            "../../bazel-out/k8-fastbuild/bin/path/to",
+            "../../bazel-out/x64_windows-fastbuild/bin/path/to",
+            "../../bazel-out/darwin-dbg/bin/path/to",
+            "../../bazel-out/k8-dbg/bin/path/to",
+            "../../bazel-out/x64_windows-dbg/bin/path/to",
+        ]
+    }
+    ```
+
+    If your dependencies are not installed into a workspace called `npm`, or if you want to
+    use a custom TypeScript compiler binary, you can pass `tsc = "@some_wksp//path/to:tsc_bin"`
+    to this rule to override the compiler.
+
+    Args:
+        name: The basename (no `.json` extension) of the tsconfig file that should be compiled.
+        srcs: List of labels of TypeScript source files to be provided to the compiler.
+
+            If absent, defaults to `**/*.ts` (all TypeScript files in the package).
+        deps: List of labels of other rules that produce TypeScript typings (.d.ts files)
+        extends: List of labels of tsconfig file(s) referenced in `extends` section of tsconfig.
+
+            Must include any tsconfig files "chained" by extends clauses.
+        declaration: if the `declaration` or `composite` bit are set in the tsconfig.
+            Instructs Bazel to expect a `.d.ts` output for each `.ts` source.
+        source_map: if the `sourceMap` bit is set in the tsconfig.
+            Instructs Bazel to expect a `.js.map` output for each `.ts` source.
+        declaration_map: if the `declarationMap` bit is set in the tsconfig.
+            Instructs Bazel to expect a `.d.ts.map` output for each `.ts` source.
+        emit_declaration_only: if the `emitDeclarationOnly` bit is set in the tsconfig.
+            Instructs Bazel *not* to expect `.js` outputs for `.ts` sources.
+    """
+
+    if srcs == None:
+        srcs = native.glob(["**/*.ts"])
+    tsconfig = name + ".json"
```

Comment:
Not sure if im a fan of this.
Could we possibly change the logic to default to `tsconfig = ctx.testonly ? 'tsconfig-test.json' : 'tsconfig.json'`.
But override with the tsconfig speciffied by the user if exists.
Usually we'd have libs where their "primary target" is a `ts_library` and this will use the same name as the dir, making it easier to reference.
This here prevents using a name that's different from your tsconfig, right?
Is there another reason why this might be a good idea?

Reply:
I'll add a comment in the code explaining this.

```
Bazel's `name` property is arbitrarily controlled by the user. But here we constrain it to match the tsconfig file.
The reason for doing this is so that BUILD file generation can be trivial. If you know the path of a referenced tsconfig file, you know what label to use to include it in the deps of another rule.
If we allowed users to control it, then BUILD file generation would need heuristic semantics to do bazel query and try to figure out which label gives a target that is usable as a dep.
```

And on the rule docs
```
You can use a Bazel alias() to give an additional label for this typescript project compilation
```

Reply:
Also note you could import the raw `ts_project` rule from the `internal` path, which I think we should leave undocumented, but it gives you the escape valve to avoid this macro logic.

Reply:
Ah yeah I see the constraint here, sounds good to me with the extra docs and code gen considered

Reply:
Discussed with @gregmagolan - even if users can control `tsconfig` and `name` attributes independently, it's still pretty easy for the generator to print all `ts_project` rules in the referenced package to find the right one. So `tsconfig` is now independent (though we recommend naming it the basename of the tsconfig.json and that default is convenient)